crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4184,6 +4193,117 @@ impl<'db> TypeInferenceBuilder<'db> {
         Ok(builder.build())
     }
 
+    fn infer_type_api_special_form(
```

Comment:
Nit: I would expect this method to be located down near the other type-expression inference methods (that is, near where it's used.)

And similar to other comments, I would rather see unified/integrated slice/arity handling for all of our special forms, rather than special handling just for the red-knot-specific ones. They are all just bracketed special forms; whether they are red-knot specific or not should not be fundamental to their implementation.